protocol.md

```diff
@@ -870,6 +902,12 @@ export interface TextDocumentClientCapabilities {
 ```typescript
 interface ClientCapabilities {
 	/**
+	 * The client supports receiving the result solely through $/partialResult notifications for requests from the client to the server.
+	 * If true, the server should send all data through $/partialResult and send null as the result in the final response.
```

Comment:
>  I think the vscode-languageclient could easily transform non-streamed responses into streamed ones and transform streamed responses into non-streamed. It could even provide and support both APIs without breaking changes.

Of course this is possible, but I think this is independent of the implementation detail whether LSP is using partialResult for all requests or not.

> It is true that the JSON RPC spec doesn't say anywhere that messages must be ordered, but it doesn't say anywhere that it is allowed to mangle them either.

This is the same thing as saying that it is undefined, which means there are no guarantees. You are reading into JSON RPC how you want it to work and not what the actual spec says.

> having them ordered is the only way that makes sense and that's probably the reason they didn't spec this, it's the only logical way.

It isn't the only way. In fact, I would expect the opposite from a modern protocol.

If a client sends a request A and a request B, and the server finished processing request B first, there is nothing about JSON RPC that prevents the server from sending the response for request B before the response for request A.

I think JSON RPC didn't spec ordering because they intended there to be no guarantees about ordering.

> If the client says call A, then B, but the JSON RPC library will sometimes tell the server the client called B, then A, that is a critical bug in the JSON RPC library.

There is nothing in the JSON RPC spec that you could point to as evidence that this is a bug.
JSON RPC does not define an ordering on requests so `A()` followed by `B()` does not establish a causal ordering. If B depends on A, then the application should not send B until it receives an acknowledgement of A.

Furthermore, even if JSON RPC did guarantee in-order delivery of requests and notifications (it doesn't), this still wouldn't be good enough for our use case because that doesn't guarantee that the server actually handled A before B. If B needs to "happen after" A then the client needs to wait for the server to acknowledge A before sending B.

> And LSP already relies on this.

I just discovered that LSP defines its own ordering here (it doesn't rely on ordering defined by JSON RPC because JSON RPC doesn't define ordering):
https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#request-notification-and-response-ordering

This means that any LSP client or server needs to take extra steps to ensure ordering on top of a generic JSON RPC implementation.

@keegancsmith it sounds like our Go implementation does not enforce this ordering? If not, we should fix. The server shouldn't have to have any timing hacks, just send notifications as fast as possible.

The LSP ordering spec isn't very explicit, but it does seem to imply that everything (requests, responses, notifications) should be serialized so we shouldn't have to worry about out-of-order.

Soo.....I think this proposal is fine from a concurrency perspective, and given that I think it is ideal to require the server to not send any data in the response for streaming requests.